src/components/Canvas/GridLayer.tsx

```diff
@@ -422,9 +477,154 @@ export default class GridLayer extends BaseLayer {
+    //   correctedLefTopScreenPoint.y +
+    //     this.rowCount * squareLength +
+    //     offsetFromPixelCanvas * 2,
+    // );
+    // ctx.lineTo(
+    //   correctedLefTopScreenPoint.x - offsetFromPixelCanvas * 2,
+    //   correctedLefTopScreenPoint.y - offsetFromPixelCanvas * 2,
+    // );
+    // ctx.stroke();
+    // ctx.restore();
+  }
+
+  drawCircleButtonAlongDashedLine(correctedLefTopScreenPoint: Coord) {
+    const offsetFromPixelCanvas =
+      DashedLineOffsetFromPixelCanvas * this.panZoom.scale;
+    const ctx = this.ctx;
+    const squareLength = this.gridSquareLength * this.panZoom.scale;
+    const leftTopCorner = {
+      x: correctedLefTopScreenPoint.x - offsetFromPixelCanvas,
+      y: correctedLefTopScreenPoint.y - offsetFromPixelCanvas,
+    };
+    const radius = ExtensionGuideCircleRadius * this.panZoom.scale;
+    const topMiddle = {
+      x: correctedLefTopScreenPoint.x + (this.columnCount * squareLength) / 2,
+      y: correctedLefTopScreenPoint.y - offsetFromPixelCanvas,
+    };
+    const rightTopCorner = {
+      x:
+        correctedLefTopScreenPoint.x +
+        this.columnCount * squareLength +
+        offsetFromPixelCanvas,
+      y: correctedLefTopScreenPoint.y - offsetFromPixelCanvas,
+    };
+    const rightMiddle = {
+      x:
+        correctedLefTopScreenPoint.x +
+        this.columnCount * squareLength +
+        offsetFromPixelCanvas,
+      y: correctedLefTopScreenPoint.y + (this.rowCount * squareLength) / 2,
+    };
+    const rightBottomCorner = {
+      x:
+        correctedLefTopScreenPoint.x +
+        this.columnCount * squareLength +
+        offsetFromPixelCanvas,
+      y:
+        correctedLefTopScreenPoint.y +
+        this.rowCount * squareLength +
+        offsetFromPixelCanvas,
+    };
+    const bottomMiddle = {
+      x: correctedLefTopScreenPoint.x + (this.columnCount * squareLength) / 2,
+      y:
+        correctedLefTopScreenPoint.y +
+        this.rowCount * squareLength +
+        offsetFromPixelCanvas,
+    };
+    const leftBottomCorner = {
+      x: correctedLefTopScreenPoint.x - offsetFromPixelCanvas,
+      y:
+        correctedLefTopScreenPoint.y +
+        this.rowCount * squareLength +
+        offsetFromPixelCanvas,
+    };
+    const leftMiddle = {
+      x: correctedLefTopScreenPoint.x - offsetFromPixelCanvas,
+      y: correctedLefTopScreenPoint.y + (this.rowCount * squareLength) / 2,
+    };
+    drawCircle(ctx, leftTopCorner, radius, "#ffffff", "#5A7FF7", 1);
+    drawCircle(ctx, topMiddle, radius, "#ffffff", "#5A7FF7", 1);
+    drawCircle(ctx, rightTopCorner, radius, "#ffffff", "#5A7FF7", 1);
+    drawCircle(ctx, rightMiddle, radius, "#ffffff", "#5A7FF7", 1);
+    drawCircle(ctx, rightBottomCorner, radius, "#ffffff", "#5A7FF7", 1);
+    drawCircle(ctx, bottomMiddle, radius, "#ffffff", "#5A7FF7", 1);
+    drawCircle(ctx, leftBottomCorner, radius, "#ffffff", "#5A7FF7", 1);
+    drawCircle(ctx, leftMiddle, radius, "#ffffff", "#5A7FF7", 1);
```

Comment:
```suggestion
  [leftTopCorner, topMiddle, rightTopCorner, rightMiddle, rightBottomCorner, bottomMiddle, leftBottomCorner, leftMiddle].forEach((position) => drawCircle(ctx, position, radius,"#ffffff", "#5A7FF7", 1 ))
```
p5) just syntax sugar